pkg/monitortests/node/watchpods/cache_check.go

```diff
@@ -0,0 +1,81 @@
+package watchpods
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func checkCacheState(ctx context.Context, kubeClient kubernetes.Interface, podInformer coreinformers.PodInformer) ([]string, error) {
+	// We are running into problems where the cached list of pods contains pods that don't exist in the kube-apiserver.
+	// Let's try to check exactly.
+	cachedPods, err := podInformer.Lister().List(labels.Everything())
+	if err != nil {
+		return nil, fmt.Errorf("error listing cached pods: %w", err)
+	}
+
+	return checkCacheStateFromList(ctx, kubeClient, cachedPods)
+}
+
+func checkCacheStateFromList(ctx context.Context, kubeClient kubernetes.Interface, cachedPods []*corev1.Pod) ([]string, error) {
+	if len(cachedPods) == 0 {
+		return nil, fmt.Errorf("somehow the lister is missing pods")
+	}
+
+	biggestResourceVersion := -1
+	for _, cachedPod := range cachedPods {
+		if len(cachedPod.ResourceVersion) == 0 {
+			return nil, fmt.Errorf("encounted cached pod/%s -n %s without a resourceversion", cachedPod.Namespace, cachedPod.Name)
+		}
+		currRV, err := strconv.Atoi(cachedPod.ResourceVersion)
+		if err != nil {
+			return nil, fmt.Errorf("cached pod/%s -n %s has a non-integer RV %q: %w", cachedPod.Namespace, cachedPod.Name, cachedPod.ResourceVersion, err)
+		}
+		if currRV > biggestResourceVersion {
+			biggestResourceVersion = currRV
```

Comment:
so it looks like you are trying to get the highest RV of all pods, and then you list in the kubeclient with that RV. Is this list RV always equivalent to the highest RV of all pods?

Reply:
Looks like this is not always the case in MicroShift. Some pods may be at resourceVersion 0 (which is invalid for an exact version match) and others may be lower than the biggest one.
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/29115/pull-ci-openshift-origin-master-e2e-aws-ovn-microshift-serial/1836509395513511936 (0 example)
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/29115/pull-ci-openshift-origin-master-e2e-aws-ovn-microshift-serial/1836559313703800832 (lower resource version example)

Reply:
I agree with the comments above. To correct this code, we would need to obtain the lastSyncedResourceVersion from the informer, stop it from receiving new updates, and then make a live list request. Otherwise, we are comparing different snapshots. Note that the list's Resource Version is not the same as the RV of individual pods/objects.

Reply:
> Some pods may be at resourceVersion 0

Pods can never be at resourceversion zero.